### PR TITLE
[RLlib] Fix test_repro_ppo learner when learner API is enabled

### PR DESCRIPTION
**Pull request by kouroshHakha**
## Why are these changes needed?
A couple of fixes that allows us to run the ppo-repro-tests with enable_learner=True
1. It makes some of the error messages better
2. It modifies the compile_results output to match the old stack better, so that the random fields that tests are built on doesn't break.

```
results = 
{
	"p1": {
		"learner_stats": {
			"total_loss" = ...,
			"mean_policy_loss" = ..., 
		}
	},
	"p2": {

	},
	"__all__": {
		"learner_stats": {
			"total_loss": ...,
			"mean_gradients": ...,
		}
                 "num_agent_steps_trained": ...,
                 "num_env_steps_trained": ....
	}
}
```
3. It makes sure the GPU resource allocation is properly handled from trainable resource allocation. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -1068,7 +1068,7 @@ py_test(
 )
 py_test(
     name = "test_impala_off_policyness",
-    tags = ["team:rllib", "algorithms_dir", "multi_gpu", "exclusive"],
```
Comment (kouroshHakha) [inline rllib/BUILD:1071]: This test is not multi-gpu test.

File: rllib/algorithms/impala/tests/test_impala_off_policyness.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 class TestIMPALAOffPolicyNess(unittest.TestCase):
     @classmethod
     def setUpClass(cls) -> None:
-        ray.init(num_gpus=0)
```
Comment (kouroshHakha) [inline rllib/algorithms/impala/tests/test_impala_off_policyness.py:19]: see the num_gpus was set to 0.

**Comment by kouroshHakha** [comment]
Ok the test passes after enabling learner_api
image

Reverting the enable to merge.

**Comment by kouroshHakha** [comment]
Pre-requisite is #32868

File: rllib/algorithms/impala/impala.py [hunk 3]
```diff
@@ -664,7 +665,7 @@ def training_step(self) -> ResultDict:
         with self._timers[SYNCH_WORKER_WEIGHTS_TIMER]:
             if self.config._enable_learner_api:
                 if train_results:
-                    pids = list(train_results["learner"]["loss"].keys())
+                    pids = list(set(train_results.keys()) - {ALL_MODULES})
```
Comment (gjoliver) [inline rllib/algorithms/impala/impala.py:668]: this feels like dark magic, is there any test or assert or something to make sure that nobody would add random additional keys to train_results?

File: rllib/algorithms/ppo/ppo.py [hunk 4]
```diff
@@ -456,7 +455,7 @@ def training_step(self) -> ResultDict:
                 # TODO (Kourosh): Train results don't match the old format. The thing
                 # that used to be under `kl` is now under `mean_kl_loss`. Fix this. Do
                 # we need get here?
-                pid: train_results["loss"][pid].get("mean_kl_loss")
+                pid: train_results[pid][LEARNER_STATS_KEY].get("mean_kl_loss")
```
Comment (gjoliver) [inline rllib/algorithms/ppo/ppo.py:458]: If I want to nit picking, this change is actually changing a public external spec.
for example, people's W&B charts, or analysis scripts will break once we change the key.
we of course should do this, but it's gonna be messy to keep the change behind a flag right?

File: rllib/algorithms/ppo/tests/test_ppo_learner.py [hunk 5]
```diff
@@ -110,7 +111,7 @@ def test_loss(self):
             learner_group.set_weights(trainer.get_weights())
             results = learner_group.update(train_batch.as_multi_agent())
 
-            learner_group_loss = results["loss"]["total_loss"]
+            learner_group_loss = results[ALL_MODULES]["total_loss"]
```
Comment (gjoliver) [inline rllib/algorithms/ppo/tests/test_ppo_learner.py:114]: btw, should this be results[ALL_MODULE]["loss"]["total_loss"] instead?

**Review by gjoliver — COMMENTED** [review]
a few nit picking questions.

File: rllib/algorithms/ppo/ppo.py (hunk 4 above)
Comment (kouroshHakha) [reply rllib/algorithms/ppo/ppo.py:458]: we were breaking it before. Now it's back to the agreement we had on the old stack.

File: rllib/algorithms/ppo/tests/test_ppo_learner.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/algorithms/ppo/tests/test_ppo_learner.py:114]: no loss key is removed. We used to have loss key as a hierarchy which was solely introduced in the Learner API which was inconsistent with the old API anyways.

File: rllib/algorithms/impala/impala.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/algorithms/impala/impala.py:668]: good point. I'll assert something in the learner.

**Review by gjoliver — APPROVED** [review]
